cpp/src/arrow/compute/exec/union_node.cc

```diff
@@ -107,27 +106,34 @@ class UnionNode : public ExecNode {
     if (input_count_.Increment()) {
       outputs_[0]->InputFinished(this, total_batches_.load());
       if (batch_count_.SetTotal(total_batches_.load())) {
-        END_SPAN(span);
         finished_.MarkFinished();
       }
     }
   }
 
   Status StartProducing() override {
-    START_SPAN(span, kind_name(), {{"node", ToString()}});
+    START_SPAN(span_, std::string(kind_name()) + ":" + label(),
+               {{"node.label", label()},
+                {"node.detail", ToString()},
+                {"node.kind", kind_name()}});
     finished_ = Future<>::Make();
+#ifdef ARROW_WITH_OPENTELEMETRY
+    finished_.AddCallback([this](const Status& st) {
```

Comment:
I know I suggested this, but there is one concern here: once finished_ completes, the node may be destroyed. (Once all nodes finish, the plan finishes, and then the plan consumer might also destroy the plan since it's done.) Also, the order in which callbacks is run is not defined. It might be safest to copy `span_` into the span closure instead of capturing `this`. (Or, you could do `finished_ = finished_.Then(...)` which would mean that this callback gets to run before others, which would ensure it runs before anything that could potentially destroy the plan.)